content/python_files/imbalance_correction.py

```diff
@@ -0,0 +1,853 @@
+        [1 - true_positive_proba[:, np.newaxis], true_positive_proba[:, np.newaxis]]
+    )
+
+    # create pandas data structures for convenience
+    X = pd.DataFrame(X, columns=[f"feature_{i}" for i in range(n_features)])
+    y = pd.Series(y, name="target")
+    return X, y, true_proba
+
+
+# %% [markdown]
+#
+# The world can generate a lot of data from a fixed stationary process. Some of that data
+# cannot be accessed at the time of the study but we generate it anyway to be able to
+# compute metrics on the future population.
+
+# %%
+n_samples = 3_000_000
+X_past, y_past, true_proba_past = sample_from_linear_model(
+    true_coef, true_intercept, n_samples, seed=0
+)
+X_future, y_future, true_proba_future = sample_from_linear_model(
+    true_coef, true_intercept, n_samples, seed=1
+)
+
+# %%
+X_past.memory_usage().sum() / 1e6  # in MB
+
+# %%
+y_past.sum()
+
+# %%
+true_positive_rate_past = y_past.mean()
+true_positive_rate_past
+
+# %% [markdown]
+#
+# Let us start this study by cheating and compute the metrics obtained from the
+# true probabilities computed from the data generating process for the future
+# feature values.
+
+# %%
+true_positive_rate_future = y_future.mean()
+true_positive_rate_future
+
+# %%
+pd.Series(true_proba_future[:, 1]).describe()
+
+# %%
+from sklearn.metrics import roc_auc_score, log_loss
+
+roc_auc_score(y_future, true_proba_future[:, 1])
+
+# %%
+log_loss(y_future, true_proba_future)
+
+# %%
+log_loss(y_past, true_proba_past)
+
+# %% [markdown]
+#
+# ### Questions:
+# - What is the best possible value for the ROC AUC score (in general)?
+# - What is the best possible value for the log-loss (in general)?
+# - Why cannot we reach these values even when using the true probabilities?
+# - What is the name name of the error rate obtained when computed from the
+#   true probabilities?
+
+# %%
+# TODO: write your answers here before scrolling down.
+#
+#
+#
+#
+#
+#
+#
+#
+#
+#
+#
+#
+
+# %% [markdown]
+#
+# ### Answers:
+# - The best possible value for the ROC AUC score is 1.0.
+# - The best possible value for the log-loss is 0.0.
+# - We cannot reach these values because the true probabilities reflect some
+#   inherent uncertainty in the data: the value of the target variable is not
+#   deterministic given the available features. The outcome can also be
+#   influenced by unobserved factors that are not captured by the features.
+# - The error rate obtained from the true probabilities is called the "Bayes
+#   error rate" or "irreducible error rate".
+
+
+# %% [markdown]
+#
+# Because we are curious, let's continue with the cheating and check that a fit
+# of logistic regression on the full population (with nearly unlimited data
+# points) would be able to approximately recover the true coefficients and
+# intercept.
+
+# %%
+from sklearn.linear_model import LogisticRegression
+
+cheating_model = LogisticRegression(penalty=None).fit(X_future, y_future)
+
+
+# %%
+class ModelComparator:
```

Comment:
It is a nice class. I'm wondering if we should have one single documentation string for each method?

Reply:
I'll add docstring later.